subprojects/core/src/main/java/org/gradle/execution/taskgraph/DefaultTaskExecutionPlan.java

```diff
@@ -538,10 +539,16 @@ public boolean executeWithTask(WorkerLease parentWorkerLease, final Action<TaskI
     }
 
     private ResourceLock getProjectLock(TaskInfo taskInfo) {
+        if (projectLocks.containsKey(taskInfo)) {
```

Comment:
After our conversation, I looked at this again and realized that it was being called in two places - one that was thread-safe, and one that wasn't.  I've changed this to a concurrent map.